### DISCUSSION TITLE
How to enable TypeScript `dom` lib via triple slash directives in `.js` files

### DISCUSSION THREAD

**Discussion by jaydenseric** [denoland/deno#11973] [Q&A] [answered]
I have a significant amount of isomorphic `.js` files that I intend to publish in a library that have have type checking enabled via a `// @ts-check` comment and JSDoc types throughout the code. This code uses the DOM (in React component hooks and event handlers, so it's ok during Deno SSR). Unfortunately, there doesn't seem to be a way to enable the built-in TypeScript `dom` lib for types within `.js` files, using the triple slash directive `/// <reference lib="dom" />`:

<img width="961" alt="Screen Shot 2021-09-10 at 3 13 59 pm" src="https://user-images.githubusercontent.com/1754873/132804324-76da51fb-28aa-47e9-b43f-e8104499dce2.png">

The Deno lint error being:

```
This triple-slash reference directive is not allowed in JavaScript
In JavaScript only the `types` directive is allowed, like `/// <reference types="..." />`
```

Why does this limitation exist? Can it be lifted?

Using [a `lib` property in a `tsconfig.json` file](https://deno.land/manual@v1.13.2/typescript/configuration#targeting-deno-and-the-browser) is not a good solution, as only some files in the project are isomorphic, and these files will be published and deep imported by users so all types information needs to be contained within the file.

I considered `/// <reference types="" />` and using a URL to a third-party Deno-compatible DOM `.d.ts` file, but this sucks because Deno ships with official DOM types so surely there is a way to reference them. It would be bad for my DOM types to conflict in user's projects that use the proper built-in DOM lib.

Sometimes it feels like I'm the only one trying to use Deno to write isomorphic, published JS (see also https://github.com/denoland/deno/issues/11362#issuecomment-909748559). I'm getting tired and desperate trying to solve this, so any help is greatly appreciated.

**Comment by kitsonk** [comment, marked as answer]
This is really just a wrong lint error. We shouldn't error on this. Because it isn't an issue, I can't transfer it to `deno_lint`. We were just too conservative in that lint rule.

**Reply by jaydenseric** [reply]
Thanks, I'll raise an issue there. This appears to work in the meantime:

```js
// @ts-check
// deno-lint-ignore no-invalid-triple-slash-reference
/// <reference lib="dom" />
// deno-lint-ignore no-invalid-triple-slash-reference
/// <reference lib="deno.ns" />
```

**Reply by jaydenseric** [reply]
Done: https://github.com/denoland/deno_lint/issues/854 .
